src/Common/ReplicasReconnector.h

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include <Core/ServerSettings.h>
+#include <Core/BackgroundSchedulePool.h>
+#include <Interpreters/Context.h>
+#include <boost/noncopyable.hpp>
+#include <functional>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <utility>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NOT_INITIALIZED;
+    extern const int LOGICAL_ERROR;
+}
+
+namespace ServerSetting
+{
+    extern const ServerSettingsUInt64 bg_reconnect_mysql_dict_interval;
+}
+
+class ReplicasReconnector : private boost::noncopyable
+{
+public:
+    using Reconnector = std::function<bool(UInt64)>;
+    using ReconnectorsList = std::list<Reconnector>;
+
+    ReplicasReconnector(const ReplicasReconnector &) = delete;
+
+    ~ReplicasReconnector()
+    {
+        emergency_stop = true;
+        task_handle->deactivate();
+        instance_ptr(nullptr, true);
+    }
+
+    [[nodiscard]]
+    static std::unique_ptr<ReplicasReconnector> init(ContextPtr context)
+    {
+        std::unique_ptr<ReplicasReconnector> ret(new ReplicasReconnector(context));
+        instance_ptr(ret.get());
+        return ret;
+    }
+
+    static ReplicasReconnector & instance()
+    {
+        ReplicasReconnector * ptr = instance_ptr();
+        if (!ptr)
+            throw Exception(ErrorCodes::NOT_INITIALIZED, "Replicas reconnector is not initialized.");
+        return *ptr;
+    }
+
+    void add(const Reconnector & reconnector)
+    {
+        std::lock_guard lock(mutex);
+        reconnectors.push_back(reconnector);
+        task_handle->activateAndSchedule();
+    }
+
+private:
+    ReconnectorsList reconnectors;
+    std::mutex mutex;
+    std::atomic_bool emergency_stop{false};
+    BackgroundSchedulePoolTaskHolder task_handle;
+
+    static ReplicasReconnector * instance_ptr(ReplicasReconnector * ptr = nullptr, bool deinitialize = false)
+    {
+        static ReplicasReconnector * instance_ptr = nullptr;
+
+        if (deinitialize)
+        {
+            instance_ptr = nullptr;
+            return instance_ptr;
+        }
```

Comment:
We can make is a class field and move the deinitialization to a separate method.

Reply:
It's impossible - the singleton holder is a static variable in this exact function - at line 74 - `static ReplicasReconnector * instance_ptr = nullptr;` - all manipulations with it can be only performed in this function. And this function is private anyway so never called directly outside, and it serves 3 functions - singleton initialization, deinitialization and access.

Reply:
It seems more convenient to keep it centralized and compact - it serves 3 functions - singleton initialization, deinitialization and access, and this function is never accessed outside... but probably having less "if's" is better and less obfuscated... I'll take a look...

Reply:
done